### ISSUE TITLE
Output ReLU Causes Training to Remain at BCE ≈ 0.6931

### ISSUE THREAD

**Issue by mahdis-saeedi** [fani-lab/OpeNTF#349]
> Hi @hosseinfani 
> While training the end-to-end GNN models, we observed that the training and validation losses remained nearly constant at approximately 0.6931, indicating that the models were not learning. Several components of the training pipeline were verified and the issue was eventually traced to the encoder. The encoder originally applied a ReLU activation after every graph convolution layer, including the final embedding layer:
> 
>                for conv in self.encoder: x = F.relu(conv(x, edge_index))
> 
> Applying ReLU to the final layer changes all negative numbers to zero, producing sparse node embeddings. In our experiments, more than 90% of the embedding entries became exactly zero. 
> 
> <img width="1058" height="341" alt="Image" src="https://github.com/user-attachments/assets/715a2f9c-0b0d-416d-9cd9-3c77f5c581f4" />
> 
> This limits the expressiveness of the prediction layer. The model predicts expert membership to a team by computing logits from the embeddings of the team and expert nodes, and these logits are optimized using BCE Loss. When the final node embeddings are highly sparse with most dimensions equal to zero, the predicted logits tend to cluster around zero, producing output probabilities close to 0.5, resulting in weak optimization signals and training that remains close to the random baseline (BCE ≈ 0.693).
> 
> 
> The encoder was modified so that ReLU is applied only to the hidden layers, leaving the final embedding layer linear:
> 
>            
>                 for i, conv in enumerate(self.encoder):
>                     x = conv(x, edge_index)
>                     if i < len(self.encoder) - 1: x = self.torch.nn.functional.relu(x)
> 
> This preserves nonlinearity throughout the hidden representation learning while allowing the final embeddings to contain both positive and negative values. As a result embedding sparsity was dramatically reduced, training and validation losses decreased normally.
> 
> <img width="1151" height="330" alt="Image" src="https://github.com/user-attachments/assets/82538dbf-9706-4775-a1f2-df64624b294c" />
> 
> <img width="388" height="495" alt="Image" src="https://github.com/user-attachments/assets/7590c480-7ad8-42ea-8561-62e315d5e2eb" />
> 
>  

 _Originally posted by @mahdis-saeedi in [#305](https://github.com/fani-lab/OpeNTF/issues/305#issuecomment-5138804869)_

**Comment by hosseinfani** [comment]
Hi @mahdis-saeedi 
Very good finding. Can you please study the effect of removing the relu from other layers, also other activation functions suited for graph embeddings, and then send me pr for the code change? Thank you.

**Comment by mahdis-saeedi** [comment]
Hi @hosseinfani .
The following table summarizes the results obtained using different activation functions for e2e experiments.

The activation function is not applied to the final layer, in order to preserve the importance of negative logits before the dot-product decoder.

Overall, the results are very similar across all activation functions. However, I notice that tanh appears to produce slightly more favorable results, as indicated by the greater number of green cells in the table.

These experiments were conducted using 2-hop neighborhood sampling (i.e., a single hidden/message-passing layer). Do you think it would be worthwhile to repeat the experiments with 3-hop neighborhood sampling to determine whether the effect of the activation function becomes more pronounced with a deeper gnn? Thank you.

<img width="1003" height="472" alt="Image" src="https://github.com/user-attachments/assets/c79e71c2-9aaa-43a4-aba2-cf3c178f3ad2" />

**Comment by hosseinfani** [comment]
Hi @mahdis-saeedi 
Generally, the results are the same for three digits. yes pls do for 3 hops but also for gcn. tnx.
